assembly/src/v1/parsers/stack_ops.rs

```diff
@@ -3,39 +3,141 @@ use super::{AssemblyError, Operation, Token};
+pub fn parse_padw(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    match op.num_parts() {
+        0 => return Err(AssemblyError::missing_param(op)),
+        1 => {
+            for _ in 0..4 {
+                span_ops.push(Operation::Pad);
+            }
+        }
+        _ => return Err(AssemblyError::extra_param(op)),
+    }
+
+    Ok(())
 }
 
-/// TODO: implement
-pub fn parse_dup(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
+/// Translates dup.n assembly instruction to VM operations DUPN.
+pub fn parse_dup(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    match op.num_parts() {
+        0 => return Err(AssemblyError::missing_param(op)),
+        1 => span_ops.push(Operation::Dup0),
+        2 => match op.parts()[1] {
+            "0" => span_ops.push(Operation::Dup0),
+            "1" => span_ops.push(Operation::Dup1),
+            "2" => span_ops.push(Operation::Dup2),
+            "3" => span_ops.push(Operation::Dup3),
+            "4" => span_ops.push(Operation::Dup4),
+            "5" => span_ops.push(Operation::Dup5),
+            "6" => span_ops.push(Operation::Dup6),
+            "7" => span_ops.push(Operation::Dup7),
+            "8" => span_ops.push(Operation::Dup8),
+            "9" => span_ops.push(Operation::Dup9),
+            "10" => span_ops.push(Operation::Dup10),
+            "11" => span_ops.push(Operation::Dup11),
+            "12" => span_ops.push(Operation::Dup12),
+            "13" => span_ops.push(Operation::Dup13),
+            "14" => span_ops.push(Operation::Dup14),
+            "15" => span_ops.push(Operation::Dup15),
+            _ => return Err(AssemblyError::invalid_param(op, 1)),
+        },
+        _ => return Err(AssemblyError::extra_param(op)),
+    }
+
+    Ok(())
 }
 
-/// TODO: implement
-pub fn parse_dupw(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
+/// Translates dupw.n assembly instruction to four VM operations DUP depending on
+/// the index of the word.
+pub fn parse_dupw(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    match op.num_parts() {
+        0 => return Err(AssemblyError::missing_param(op)),
+        1 => {
+            span_ops.push(Operation::Dup3);
+            span_ops.push(Operation::Dup2);
+            span_ops.push(Operation::Dup1);
+            span_ops.push(Operation::Dup0);
+        }
+        2 => match op.parts()[1] {
+            "0" => {
+                span_ops.push(Operation::Dup3);
+                span_ops.push(Operation::Dup2);
+                span_ops.push(Operation::Dup1);
+                span_ops.push(Operation::Dup0);
+            }
+            "1" => {
+                span_ops.push(Operation::Dup7);
+                span_ops.push(Operation::Dup6);
+                span_ops.push(Operation::Dup5);
+                span_ops.push(Operation::Dup4);
```

Comment:
Shouldn't this be: `DUP7 DUP7 DUP7 DUP7`?

Reply:
ah yes you're right! Fixed all the other places